### PR TITLE
add type annotation in io module

### PR DESCRIPTION
**Pull request by WassCodeur**
Hello @skoudoro! I hope you are well. 

I've been working on type annotation in the io modules (issue #2803) please could you check this and give me feedback if possible ?

### CODE REVIEW

File: dipy/io/dpy.py [hunk 1]
```diff
@@ -85,12 +87,24 @@ def __init__(self, fname, mode='r', compression=0):
...
-    def version(self):
+    def version(self) -> str:
 
         return self.f.attrs['version']
 
-    def write_track(self, track):
+    def write_track(self, track: np.array) -> None:
         """ write on track each time
+
+        Args
```
Comment (skoudoro) [inline dipy/io/dpy.py:97]: Can you follow numpy documentation standard : https://numpydoc.readthedocs.io/en/latest/format.html

File: dipy/io/dpy.py [hunk 2]
```diff
@@ -99,8 +113,15 @@ def write_track(self, track):
         self.offsets.resize(self.offsets.shape[0] + 1, axis=0)
         self.offsets[-1] = self.curr_pos
 
-    def write_tracks(self, tracks):
+    def write_tracks(self, tracks: Streamlines) -> None:
```
Comment (skoudoro) [inline dipy/io/dpy.py:116]: same comment as above:

Can you follow numpy documentation standard : https://numpydoc.readthedocs.io/en/latest/format.html

File: dipy/io/dpy.py [hunk 3]
```diff
@@ -113,34 +134,54 @@ def write_tracks(self, tracks):
...
         """
         off0, off1 = self.offsets[self.offs_pos:self.offs_pos + 2]
         self.offs_pos += 1
         return self.tracks[off0:off1]
 
-    def read_tracksi(self, indices):
+    def read_tracksi(self, indices: List[int]) -> Streamlines:
         """ read tracks with specific indices
+        Args:
+                indices: a list of indices
```
Comment (skoudoro) [inline dipy/io/dpy.py:151]: Can you follow numpy documentation standard : https://numpydoc.readthedocs.io/en/latest/format.html

File: dipy/io/dpy.py [hunk 4]
```diff
@@ -113,34 +134,54 @@ def write_tracks(self, tracks):
...
         tracks = Streamlines()
         for i in indices:
             off0, off1 = self.offsets[i:i + 2]
             tracks.append(self.tracks[off0:off1])
         return tracks
 
-    def read_tracks(self):
+    def read_tracks(self) -> Streamlines:
         """ read the entire tractography
+        Args:
```
Comment (skoudoro) [inline dipy/io/dpy.py:163]: Can you follow numpy documentation standard : https://numpydoc.readthedocs.io/en/latest/format.html

**Review by skoudoro — COMMENTED** [review]
Thank you for this @WassCodeur!

I will look more deeply into it a bit later today. 

For now, can you fix all docstring issue ? Thank you

File: dipy/io/dpy.py (hunk 1 above)
Comment (WassCodeur) [reply dipy/io/dpy.py:97]: Okay

File: dipy/io/dpy.py (hunk 4 above)
Comment (WassCodeur) [reply dipy/io/dpy.py:163]: Okay

File: dipy/io/dpy.py (hunk 2 above)
Comment (WassCodeur) [reply dipy/io/dpy.py:116]: okay

File: dipy/io/dpy.py [hunk 5]
```diff
@@ -12,13 +12,15 @@
 import h5py
 
 from nibabel.streamlines import ArraySequence as Streamlines
+from typing import List
 
 # Make sure not to carry across setup module from * import
 __all__ = ['Dpy']
 
 
 class Dpy:
```
Comment (WassCodeur) [inline dipy/io/dpy.py:21]: Missing class docstring

File: dipy/io/dpy.py [hunk 6]
```diff
@@ -85,12 +87,24 @@ def __init__(self, fname, mode='r', compression=0):
             self.track_no = len(self.offsets) - 1
             self.offs_pos = 0
 
-    def version(self):
+    def version(self) -> str:
```
Comment (WassCodeur) [inline dipy/io/dpy.py:90]: add method docstring

File: dipy/io/dpy.py [hunk 7]
```diff
@@ -88,23 +116,37 @@ def __init__(self, fname: str, mode: str = 'r',
             self.offs_pos = 0
 
     def version(self) -> str:
+        """
+        return the version of the file
+
+        Parameters:
```
Comment (skoudoro) [inline dipy/io/dpy.py:122]: colon `:` to remove

File: dipy/io/dpy.py [hunk 8]
```diff
@@ -88,23 +116,37 @@ def __init__(self, fname: str, mode: str = 'r',
...
+        -----------
+        None
 
+        Returns:
```
Comment (skoudoro) [inline dipy/io/dpy.py:126]: colon `:` to remove

File: dipy/io/dpy.py [hunk 9]
```diff
@@ -114,14 +156,22 @@ def write_track(self, track: np.array) -> None:
         self.offsets[-1] = self.curr_pos
 
     def write_tracks(self, tracks: Streamlines) -> None:
-        """ write many tracks together
-        Args:
-                tracks: a Streamlines object
+        """
+        write many tracks together
+
+        Parameters:
```
Comment (skoudoro) [inline dipy/io/dpy.py:162]: colon `:` to remove

File: dipy/io/vtk.py [hunk 10]
```diff
@@ -8,7 +10,7 @@
     import fury.io
 
 
-def load_polydata(file_name):
+def load_polydata(file_name: str) -> callable:
```
Comment (skoudoro) [inline dipy/io/vtk.py:13]: it is not a callable but a polydata

File: dipy/io/vtk.py [hunk 11]
```diff
@@ -40,8 +44,9 @@ def save_polydata(polydata, file_name, binary=False, color_array_name=None):
                           binary=binary, color_array_name=color_array_name)
 
 
-def save_vtk_streamlines(streamlines, filename,
-                         to_lps=True, binary=False):
+def save_vtk_streamlines(streamlines: Union[ArraySequence, list],
```
Comment (skoudoro) [inline dipy/io/vtk.py:47]: Union[ArraySequence, list], Streamlines instead, you might need to create a type name

File: dipy/io/vtk.py [hunk 12]
```diff
@@ -70,7 +75,7 @@ def save_vtk_streamlines(streamlines, filename,
     save_polydata(polydata, file_name=filename, binary=binary)
 
 
-def load_vtk_streamlines(filename, to_lps=True):
+def load_vtk_streamlines(filename: str, to_lps: bool = True) -> list:
```
Comment (skoudoro) [inline dipy/io/vtk.py:78]: List[np.ndarray] instead ?

**Review by skoudoro — COMMENTED** [review]
Hi @wasscodeur,

Thank you for the update. See below for some additional comments.

Make sure to remove colon on the docstring and shorten the line below. 

Also, please, check a last time if all typing make sense. Thank you !

File: dipy/io/dpy.py (hunk 7 above)
Comment (WassCodeur) [reply dipy/io/dpy.py:122]: okay

File: dipy/io/dpy.py (hunk 9 above)
Comment (skoudoro) [reply dipy/io/dpy.py:162]: you forgot the colon here

File: dipy/io/dpy.py [hunk 13]
```diff
@@ -99,8 +155,23 @@ def write_track(self, track):
...
+        -----------
+        tracks: a Streamlines object
+                    tracks data to be written
+
+        Returns:
+        --------
+        None
+
+        Notes:
```
Comment (skoudoro) [inline dipy/io/dpy.py:171]: you forgot to remove the colon here

File: dipy/io/dpy.py [hunk 14]
```diff
@@ -99,8 +155,23 @@ def write_track(self, track):
...
+    def write_tracks(self, tracks: Streamlines) -> None:
+        """
+        write many tracks together
+
+        Parameters:
+        -----------
+        tracks: a Streamlines object
+                    tracks data to be written
+
+        Returns:
```
Comment (skoudoro) [inline dipy/io/dpy.py:167]: you forgot to remove the colon here

**Review by skoudoro — COMMENTED** [review]
Hi @WassCodeur,

Overall, it looks good. 

Does anyone have any additional comment? @arokem ? @jhlegarreta ?

ps: we might need to define a typing module with our custom type.

**Comment by jhlegarreta** [comment]
> Does anyone have any additional comment? @arokem ? @jhlegarreta ?

If types will be annotated in the method signature, then we should remove them from the docstrings, e.g.
```
def read_bvec_file(filename: str, atol: Optional[float] = .001):
```
at https://github.com/dipy/dipy/pull/2960/files#diff-c3d1b4f20d295179852e85468112457725c9e609cd1e0b29efceafcdc3a34b36R12, and
```
    atol : float, optional
```
at https://github.com/dipy/dipy/pull/2960/files#diff-c3d1b4f20d295179852e85468112457725c9e609cd1e0b29efceafcdc3a34b36R25

Same goes for the `optional` label that we use in the docstrings.

Defining the types in two places is hard to maintain, and we run the risk of going out of sync and having a documentation that contradicts itself.

I have not looked into how other scientific packages have solved this.

I'd refrain ourselves from merging this until a solution is found, or at least until it is investigated.

In a longer term perspective, defining parameters and their documentation in a single place would be desirable: this would save typing effort, ease maintenance, and allow to have a consistent API. I ignore what the status of this is in other major scientific packages, but I believe some already implement this (cannot spend much time finding examples now, but an example of a recommendation is https://matplotlib.org/stable/devel/MEP/MEP10.html). Involves some (a lot of) work, but pays off down the road.

**Comment by skoudoro** [comment]
After the last DIPY Open Meeting, we decided to wait and see the evolution of type annotation adoption.

So closing this PR for now. 

Sorry @WassCodeur. Thanks for your nice work. For sure it will be useful when we will restart this work.

**Pull request closed without merging**